reactivesocket-client/src/main/java/io/reactivesocket/client/LoadBalancer.java

```diff
@@ -313,38 +335,44 @@ private void updateAperture(int newValue, long now) {
      */
     private synchronized void refreshSockets() {
         refreshAperture();
-
-        int n = pendingSockets + activeSockets.size();
-        if (n < targetAperture && !activeFactories.isEmpty()) {
+        int n = pendingSockets + activeSockets.holder.size();
+        if (n < targetAperture && !activeFactories.holder.isEmpty()) {
             logger.debug("aperture {} is below target {}, adding {} sockets",
                 n, targetAperture, targetAperture - n);
             addSockets(targetAperture - n);
-        } else if (targetAperture <  activeSockets.size()) {
+        } else if (targetAperture <  activeSockets.holder.size()) {
             logger.debug("aperture {} is above target {}, quicking 1 socket",
                 n, targetAperture);
             quickSlowestRS();
         }
 
         long now = Clock.now();
-        if (now - lastRefresh < refreshPeriod) {
-            return;
+        if (now - lastRefresh >= refreshPeriod) {
+            if (eventListener != null) {
+                eventListener.socketsRefreshStart();
+            }
+            long prev = refreshPeriod;
+            refreshPeriod = (long) Math.min(refreshPeriod * 1.5, maxRefreshPeriod);
+            logger.debug("Bumping refresh period, {}->{}", prev / 1000, refreshPeriod / 1000);
+            if (prev != refreshPeriod && eventListener != null) {
+                eventListener.socketRefreshPeriodChanged(prev, refreshPeriod, Clock.unit());
+            }
+            lastRefresh = now;
+            addSockets(1);
+            if (eventListener != null) {
+                eventListener.socketsRefreshCompleted(Clock.elapsedSince(now), Clock.unit());
```

Comment:
(Discussed offline)
It is more involved to emit this event after the `addSocket` is finished. This event is more useful to indicate that socket refresh is happening, latency of connect can be useful in determining latency of refresh.